eth/downloader/queue.go

```diff
@@ -548,29 +863,101 @@ func (q *queue) Deliver(id string, txLists [][]*types.Transaction, uncleLists []
 	}
 }
 
-// enqueue inserts a new block into the final delivery queue, waiting for pickup
-// by the processor.
-func (q *queue) enqueue(origin string, block *types.Block) error {
-	// If a requested block falls out of the range, the hash chain is invalid
-	index := int(int64(block.NumberU64()) - int64(q.blockOffset))
-	if index >= len(q.blockCache) || index < 0 {
-		return errInvalidChain
+// DeliverNodeData injects a node state data retrieval response into the queue.
+func (q *queue) DeliverNodeData(id string, data [][]byte, callback func(error, int)) error {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// Short circuit if the data was never requested
+	request := q.statePendPool[id]
+	if request == nil {
+		return errNoFetchesPending
+	}
+	stateReqTimer.UpdateSince(request.Time)
+	delete(q.statePendPool, id)
+
+	// If no data was retrieved, mark them as unavailable for the origin peer
+	if len(data) == 0 {
+		for hash, _ := range request.Hashes {
+			request.Peer.ignored.Add(hash)
+		}
+	}
+	// Iterate over the downloaded data and verify each of them
+	errs := make([]error, 0)
+	process := []trie.SyncResult{}
+	for _, blob := range data {
+		// Skip any blocks that were not requested
+		hash := common.BytesToHash(crypto.Sha3(blob))
+		if _, ok := request.Hashes[hash]; !ok {
+			errs = append(errs, fmt.Errorf("non-requested state data %x", hash))
+			continue
```

Comment:
It is. https://github.com/karalabe/go-ethereum/blob/fast-sync-rebase/eth/downloader/downloader.go#L1391